### PR TITLE
Add updates concerning dynamic profile generation

### PR DESCRIPTION
**Pull request by zadjii-msft**
This is based on discussion with @dhowett-msft we had o*line. We're trying to
  work through a way to prevent dynamic profiles from roaming to machines the
  dynamic profiles might not exist on.

  After writing this up, I'm not totally sure that it's a better design.

I'm making a PR for this element of the design into the other spec. It seemed like a pretty substantial change to the other spec that should almost be reviewed on its own. If we agree on this bit of it, I'll merge it in to be reviewed with the rest.

### CODE REVIEW

File: doc/cascadia/Cascading-Default-Settings.md [hunk 1]
```diff
@@ -219,15 +229,33 @@ profiles. The generator will return some sort of result indicating that it wants
 a save operation to occur. The app will then save the `profiles.json` file,
 including these new profiles.
 
-If a dynamic profile generator has determined that a profile should no longer be
+We'll generate these dynamic profiles immediately after applying the defaults.
+When a generator runs, it'll be able to create unique profile GUIDs for each
+source it wants to generate a profile for. It'll attempt find any profiles that
```
Comment (DHowett-MSFT) [inline doc/cascadia/Cascading-Default-Settings.md]: It shouldn't attempt to find any profiles that exist -- right now, during the "early phase," there shouldn't be any profiles that exist except the ones in the defaults. The ones from the json file have been read from disk, but _not yet turned into profile objects_.

The generator will create a bunch of basics, which are then _strictly overlaid_ with the user's json json tree. The generator doesn't ever need to query the user's store.

File: doc/cascadia/Cascading-Default-Settings.md [hunk 2]
```diff
@@ -444,6 +488,39 @@ the `profiles.json` file. Fortunately though, users should be able to remove
 much of the boilerplate from their `profiles.json` files, and trim it down just
 to their modifications.
 
+### Dynamic Profile Generators Need to be Enabled
+With the current proposal, profiles that are generated by a dynamic profile
+generator _need_ that generator to be enabled for the profile to appear in the
```
Comment (DHowett-MSFT) [inline doc/cascadia/Cascading-Default-Settings.md]: This is required for things like "I uninstalled my WSL distribution, why is it still hanging around" to work, anyway. Maybe if we bang on the language a bit to change "generator" to "source"? A "profile source" can be disabled, and _that will remove all of the profiles that originated from that source._

Comment (DHowett-MSFT) [reply doc/cascadia/Cascading-Default-Settings.md]: I think the discussion around "profile generators can be slow" should be resolved by "don't do that, you are a bad plugin author" :smile:

File: doc/cascadia/Cascading-Default-Settings.md [hunk 3]
```diff
@@ -458,6 +535,15 @@ to their modifications.
   behave correctly. It's possible that we could abstract our implementation into
   a WinRT interface that extensions could implement, and be triggered just like
   other dynamic profile generators.
+  - Additionally, we could expand the dynamic profile generator contract to
```
Comment (DHowett-MSFT) [inline doc/cascadia/Cascading-Default-Settings.md]: Now here is a thought. If we _force_ the GUID, we can encode `source: '{guid-of-source}'` in the user copy of the profile. If a profile has a `source`, but that source didn't early-generate a profile of the same guid, THEN we just fail to apply. We don't need any logic to say "if the command line is blank, skip it"!

File: doc/cascadia/Cascading-Default-Settings.md [hunk 4]
```diff
@@ -474,6 +560,10 @@ to their modifications.
   that as the point of comparison to check if a setting's value has changed.
   There may be more unknowns with this proposal, so I leave it for a future
   feature spec.
+  - How do these play with dynamically generated profiles? If the user sets a
```
Comment (DHowett-MSFT) [inline doc/cascadia/Cascading-Default-Settings.md]: Above `source-guid` comment might help address this!

File: doc/cascadia/Cascading-Default-Settings.md (hunk 2 above)
Comment (DHowett-MSFT) [reply doc/cascadia/Cascading-Default-Settings.md]: This also opens us up to "default.json parser is just another type-of-source", which might be a safe mental model?

**Comment by DHowett-MSFT** [comment]
The flow becomes:

1. load `default.json` from disk -> DefaultsJson
1. load `profiles.json` from disk -> UserJson
1. parse DefaultsJson into Profile objects -> new Profile objects (1)
1. run `<all generators>::Generate` -> new Profile objects (2)
1. union all Profile objects (1) and (2) -> Profile set (3)
1. overlay UserJson.globals.defaults on all Profiles in set (3)
1. overlay UserJson.<profiles> on all Profiles in set (3)
  * If a profile with a `source` is not attested by that source (check set (2)), _we can't overlay it (!)_ so it is never created, so it never needs to be deleted.

Done. User "global defaults" override all generator settings, profile specific settings override all settings, and generators have 0 reliance on UserJson.

File: doc/cascadia/Cascading-Default-Settings.md (hunk 1 above)
Comment (zadjii-msft) [reply doc/cascadia/Cascading-Default-Settings.md]: I think there's one other step you missed. We need to make a first pass on the user's settings to make sure all the profiles end up in the right order. We'll do nothing in that pass other than instantiate a list of profiles with matching GUIDs. When we generate profiles dynamically, their settings will be applied to the profile objects that already exist, as if only to maintain the ordering.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT how about: 

When we load the settings, we'll perform the following steps, each mentioned in greater detail below:
1. Load from disk the `defaults.json` (the default settings) -> DefaultsJson
1. Load from disk the `profiles.json` (the user settings) -> UserJson
1. Perform a preliminary scan of the user settings, and create all the profiles in order they appear in the user settings file. Fill in their `guid` and `source` properties (if they exist).
1. Layer all settings from the defaults on the existing profiles, and create the default color schemes, as well as the default global settings.
   - To layer: if a profile exists already in the list of profiles with a matching `guid` and `source`, apply the settings _to that Profile object_. Otherwise, if there is no match, append it to the end of the list of profiles.
1. Generate any dynamically generated profiles, if necessary. Layer them on top of existing profiles. If any profiles were appended in this manner, indicate that we'll need to save the updated settings.
1. [Not covered in this spec] Apply the UserJson.globals.defaults settings to every profile, including default, generated and user profiles. (This will override generated settings for generated profiles).
1. Layer all user settings upon the existing settings models. 
1. Validate the settings.
1. If necessary, write the modified settings back to `profiles.json`.


I think functionally it's the same, with an added step to make sure the profiles appear in the order they appear in the user settings file

File: doc/cascadia/Cascading-Default-Settings.md (hunk 1 above)
Comment (DHowett-MSFT) [reply doc/cascadia/Cascading-Default-Settings.md]: That doesn't really matter, we can put profiles in order once we load the user's settings. We're using a vector _right now_ but that's an implementation detail.

**Comment by DHowett-MSFT** [comment]
I guess I just don't get why 4, 5 need to happen before 3. The total order can be enforced during 7, when we're "applying the user's settings". Before that, they have no order. :smile:

File: doc/cascadia/Cascading-Default-Settings.md [hunk 5]
```diff
@@ -213,31 +255,60 @@ generator runs, it can check if that profile source already has a profile
...
 profiles. The generator will return some sort of result indicating that it wants
 a save operation to occur. The app will then save the `profiles.json` file,
 including these new profiles.
 
-If a dynamic profile generator has determined that a profile should no longer be
+We'll generate these dynamic profiles immediately after applying the defaults.
+When a generator runs, it'll be able to create unique profile GUIDs for each
+source it wants to generate a profile for. It'll attempt find any profiles that
+exist in the list of profiles with a matching GUID, and apply the settings to
+that profile if found, or it'll create a new profile to apply the settings to.
```
Comment (DHowett-MSFT) [inline doc/cascadia/Cascading-Default-Settings.md]: this wording still suggests that the generator knows-about the user profile object and updates it.

File: doc/cascadia/Cascading-Default-Settings.md [hunk 6]
```diff
@@ -213,31 +255,60 @@ generator runs, it can check if that profile source already has a profile
...
+memory during runtime to be able to ensure the only state we're serializing is
+that which is different from the initially generated dynamic profile.
+
+When the generator is run, and determines that a new profile has been added,
+we'll need to make sure to add the profile to the user's settings file. This
+will create an easy point for users to customize the dynamic profiles. When
+added to the user settings, all that will be added is the `name`, `guid`, and
+`source`.
+
+<!-- If a dynamic profile generator has determined that a profile should no longer be
```
Comment (DHowett-MSFT) [inline doc/cascadia/Cascading-Default-Settings.md]: nix commented code when merging

**Pull request merged**